dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/config/ServiceConfig.java

```diff
@@ -128,6 +146,38 @@
     @Qualifier( "maxAttempts" )
     private ConfigurationPropertyFactoryBean maxAttempts;
 
+    private final Map<Class<? extends Checker>, Checker> checkersByClass;
+
+    private final Map<Class<? extends ValidationCheck>, ValidationCheck> validationCheckByClass;
+
+    private final Map<Class<? extends Processor>, Processor> processorsByClass;
+
+    private final Map<Class<? extends ProgramRuleActionValidator>, ProgramRuleActionValidator> programRuleActionValidatorsByClass;
+
+    private final Map<EventProcessorPhase, List<Processor>> processorsByPhase;
+
+    public ServiceConfig( Collection<Checker> checkers, Collection<ValidationCheck> validationChecks,
+        Collection<Processor> processors, Collection<ProgramRuleActionValidator> programRuleActionValidators )
+    {
+        checkersByClass = checkers.stream()
+            .collect( Collectors.toMap(
+                Checker::getClass,
+                Functions.identity() ) );
+        validationCheckByClass = validationChecks.stream()
+            .collect( Collectors.toMap(
+                ValidationCheck::getClass,
+                Functions.identity() ) );
+        processorsByClass = processors.stream()
+            .collect( Collectors.toMap(
+                Processor::getClass,
+                Functions.identity() ) );
+        programRuleActionValidatorsByClass = programRuleActionValidators.stream()
+            .collect( Collectors.toMap(
+                ProgramRuleActionValidator::getClass,
+                Functions.identity() ) );
```

Comment:
I think this is 4 times the same and you could even use `Object::getClass` so this does not have to become an argument to a helper function.

Reply:
> @gnespolino there are some great changes that make things a bit lighter in terms of hierarchies one has to read through/keep in mind to understand what's happening 👍🏻
> 
> You mentioned that we instantiated these checkers, validators, processors for each event being checked, validated, ... Do you have any idea how much that was costing us in terms of time/memory? It would be great if we could build our muscle of coming up with such measurements.
> 
> After reading the code for quite some time now, I wonder if could we make it even leaner? Do I understand correctly that a lot of the code we have in the `ServiceConfig` is to create collections of checkers, validators, processors, ... which are then used in some Factory which is then used in some class actually using the checkers, validators, ... on a list of events.
> 
> For example
> 
> * `validatorMap` used in  `ValidationFactory`
> * `eventInsertValidatorMap` used in `InsertValidationFactory` used in `EventManager`
> * `eventInsertPreProcessorMap` used in `PreInsertProcessorFactory` used in `ProcessingManager` used in `EventManager`
> 
> Note most of the names above are from the old code.
> 
> Since there is never any config other than `ServiceConfig` for dxf2 could we inline some of this code? What I mean by that is for example remove the code in the Service or move it to the actual Factory using the validators, checkers, ... Even better remove the factories.
> 
> Maybe use https://www.baeldung.com/spring-autowire-generics#autowiring-generics-in-spring-40 so the EventManager has its EventProcessors autowired directly? It could also be done without Spring of course, mostly with the code that is already in ServiceConfig just moved to the actual place where processors, validators, .. are used without creating these intermediary beans/fields
> 
> This is just an idea smile

It's not a bad idea, however, this code belongs to the old importer, and I don't see it as beneficial to refactor it in that way now. I think this configuration class is not that complex to understand.

Reply:
> I think this is 4 times the same and you could even use `Object::getClass` so this does not have to become an argument to a helper function.

I tried like this:
```
    private <T> Map<Class<? extends T>, T> byClass( Collection<? extends T> items )
    {
        return items.stream()
            .collect( Collectors.toMap(
                T::getClass,
                Functions.identity() ) );
    }
```

but it fails to compile with:
```
.../dhis2/dhis2-core/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/config/ServiceConfig.java:184:21
java: incompatible types: inference variable K has incompatible bounds
    equality constraints: java.lang.Class<? extends T>
    lower bounds: java.lang.Class<? extends java.lang.Object>
```

Reply:
I am surprised it lets you even do `T::getClass` and not asks for `Object::getClass`. `T::getClass` might still not get you the right type so you would need a cast. Also I think `Collection<? extends T>` as input is not mapping to `T` as output for map value when using `Functions.identity()`. Maybe this is more successful this way:
```java
private <T> Map<Class<? extends T>, T> byClass( Collection<T> items )
    {
        return items.stream()
            .collect( Collectors.toMap(
                e -> (Class<? extends T>) e.getClass(),
                Functions.identity() ) );
    }
```
You might get away with  `T::getClass` in this.

Reply:
yep this works, it's unchecked however :) But still better than repeating the same 4 times in the constructor